### PR TITLE
[Relative queries return old data] - Fixes #12314

### LINKED ISSUE #12314: [Bug]: Relative queries return old data
### Bug description

When we call `/api/v1/data` with a relative timeframe (e.g. `before=0` and `after=-600` to show the last 10 minutes), the agent responds with the relative timeframe against the last time a chart was collected.

This means that:

1. for charts that are currently being collected, the agent responds with the last 10 minutes.
2. for charts that are no longer being collected, the agent responds with 10 minutes of data, relative to the time the chart was last collected.

This makes things on the dashboard very confusing. Charts are not always aligned and if a chart failed to be updated recently the user accidentally sees stale data on the charts (the x-axis is right, but still the user needs to pay attention to notice it).

### Expected behavior

When we ask data for the last 10 minutes, the duration of the data returned should be of the last 10 minutes, independently of the whether the chart stopped being collected.

### Steps to reproduce

1. Start a container
2. Wait a few minutes for netdata to gather some data
3. Stop the container
4. Wait a few minutes
5. refresh the agent dashboard
6. the container charts do not have a gap at the end of the chart. They feel "recent".

### Installation method

kickstart.sh

### System info

```shell
N/A
```

### Netdata build info

```shell
N/A
```

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by erdem2000**
Fixes [#12314](https://github.com/netdata/netdata/issues/12314)

##### Summary
Issue:
The query returns slate data when a chart becomes invalid due to several reasons such as;
removed USB disk or stopped application services.

Root Problem:
"Before" parameter is set relatively to the "last_entry_t".

Solution:
"Before" parameter is set relatively to the "now_realtime_sec()".
"last_entry_t" checks/conditions are removed.

##### Test Plan
Expected behavior
When we ask data for the last 10 minutes, the duration of the data returned should be of the last 10 minutes, independently of the whether the chart stopped being collected.

Steps to reproduce:
- Start a container
- Wait a few minutes for netdata to gather some data
- Stop the container
- Wait a few minutes
- refresh the agent dashboard
- the container charts do not have a gap at the end of the chart. They feel "recent".

##### Additional Information
Fix is tested with steps below:
- First we need a chart which can be disabled easily to reproduce the issue, I used a USB disk.
- Plugin the USB disk to the machine which runs netdata agent and wait one minute to get metrics of `disk_busy.sdb`
- Check the metrics with `http://127.0.0.1:19999/api/v1/data?chart=disk_busy.sdb&after=-10&options=jsonwrap` request on the browser. Don't forget to change `sdb `with your USB disk name.
- Then remove the USB disk and wait 20-30 seconds.
- When you call the same request, it responds with the last 10 metrics of USB disk. But expected output is "`NULL`" because it does not get the metrics for last 20-30 seconds.
- This PR enables creating of response relative to the time of request not the last metric store time.

### CODE REVIEW

**Comment by stelfrag** [comment]
There are some checks that produce messages in the error.log like 
```
2022-03-10 12:41:46: netdata ERROR : ACLK_Query_3 : INTERNAL CHECK: before_wanted 1646908818 is too big, maximum 1646908394
```

**Comment by erdem2000** [comment]
There were extra coding lines which checks "before_wanted" and "after_wanted" parameters if they are bigger than "last_entry_t". They are also commented. 

Optionally:
We could have also left the code uncommented and changed "last_entry_t" with "now_realtime_sec()" but it wouldn't make sense because of there is not any possibility of it.

**Comment by thiagoftsm** [comment]
@erdem2000 your PR is introducing the following warning:

```sh
web/api/queries/query.c: In function 'rrd2rrdr_fixedstep':
web/api/queries/query.c:848:18: warning: unused parameter 'last_entry_t' [-Wunused-parameter]
  848 |         , time_t last_entry_t
```

please, take a look on this.

**Comment by thiagoftsm** [comment]
@erdem2000 about the description:

```
Stop the container
Wait a few minutes
refresh the agent dashboard
the container charts do not have a gap at the end of the chart. They feel "recent".
```

When we refresh dashboard when we are running only one container, we expect the chart will be removed with reload after few minutes. What do you mean  with "feel recent"?

**Comment by erdem2000** [comment]
> ```shell
> rrd2rrdr_fixedstep
> ```



> 

Actually these steps are taken from the bug, let me explain it in a detailed way.
- First we need a chart which can be disabled easily to reproduce the issue, I used a USB disk.
- Plugin the USB disk to the machine which runs netdata agent and wait one minute to get metrics of `disk_busy.sda`
- Check the metrics with `http://127.0.0.1:19999/api/v1/data?chart=disk_busy.sdb&after=-10&options=jsonwrap` request on the browser. Don't forget to change `sdb` with your USB disk name. 
- Then remove the USB disk and wait 20~30 seconds. 
- When you call the same request, it responds with the last 10 metrics of USB disk, because _**it feels it the most recent values :)**_ . But expected output is "NULL" because it does not get the metrics for last 20~30 seconds.
This PR  enables creating of response relative to the time of request not the last metric store time.

**Comment by erdem2000** [comment]
> @erdem2000 about the description:
> 
> ```
> Stop the container
> Wait a few minutes
> refresh the agent dashboard
> the container charts do not have a gap at the end of the chart. They feel "recent".
> ```
> 
> When we refresh dashboard when we are running only one container, we expect the chart will be removed with reload after few minutes. What do you mean with "feel recent"?

Actually these steps are taken from the bug, let me explain it in a detailed way.

First we need a chart which can be disabled easily to reproduce the issue, I used a USB disk.
Plugin the USB disk to the machine which runs netdata agent and wait one minute to get metrics of disk_busy.sda
Check the metrics with http://127.0.0.1:19999/api/v1/data?chart=disk_busy.sdb&after=-10&options=jsonwrap request on the browser. Don't forget to change sdb with your USB disk name.
Then remove the USB disk and wait 20~30 seconds.
When you call the same request, it responds with the last 10 metrics of USB disk, because it feels it the most recent values :) . But expected output is "NULL" because it does not get the metrics for last 20~30 seconds.
This PR enables creating of response relative to the time of request not the last metric store time.

**Review by thiagoftsm — DISMISSED** [review]
After last commit, I retest the PR and I got the expected result.
I did not observed the error that @stelfrag report in my local logs. 
LGTM!

**Comment by stelfrag** [comment]
@erdem2000 any progress?

**Comment by erdem2000** [comment]
Sorry for the late return. The last commit also fixes stationary metric output with `ram ` and `save ` memory modes.

We should also investigate the behavior below & if it is necessary open another issue for it.
When investigating the reason of this issue (with `ram `memory mode) I also figured out that `[rd->state->query_ops.next_metric(&handle, &db_now);](https://github.com/netdata/netdata/blob/b0e6c0700e6ea9615d9b2f4d4352ec8ea965eb05/web/api/queries/query.c#L593)` function returns the storage number with the `SN_ANOMALY_BIT` at the first call. 
Because of that `[does_storage_number_exist(n)](https://github.com/netdata/netdata/blob/b0e6c0700e6ea9615d9b2f4d4352ec8ea965eb05/web/api/queries/query.c#L602)` check returns `true ` and the metric is evaluated as a number instead of `null`.

If this behavior is fixed then we can revoke the related part of the PR.

**Comment by stelfrag** [comment]
@erdem2000 please rebase to resolve conflicts

**Comment by erdem2000** [comment]
> @erdem2000 please rebase to resolve conflicts

Done!

File: web/api/queries/query.c [hunk 1]
```diff
@@ -440,6 +444,12 @@ static inline void do_dimension_variablestep(
                 // use the previously read database value
                 n_curr = n_prev;
             } else {
+                if (no_db_one_time == 1 && rd->rrd_memory_mode != RRD_MEMORY_MODE_DBENGINE)
+                {
+                    no_db_one_time = 0;
+                    rd->state->query_ops.next_metric(&handle, &db_now);
+
```
Comment (thiagoftsm) [inline web/api/queries/query.c:451]: Please, remove this unnecessary blank line.
 I observed that sometimes you wrote `if () {`, and sometimes you start the `{` in the other line. The majority of our code starts the `{` in the same line of the `if`, I suggest you do the same for we keep a pattern.

File: web/api/queries/query.c [hunk 2]
```diff
@@ -826,12 +861,16 @@ static int rrdr_convert_before_after_to_absolute(
...
         absolute_period_requested = 1;
 
-    // make sure they are within our timeframe
-    if(before_requested > last_entry_t)  before_requested = last_entry_t;
+    // Make sure they are within our timeframe
+    // Timeframe can be forward than the "last_entry_t",
+    // due to invalid charts like removed USB disk or stopped application services.
+    // Therefore "last_entry_t" checks/conditions are commented.
+
+    //if(before_requested > last_entry_t)  before_requested = last_entry_t;
```
Comment (thiagoftsm) [inline web/api/queries/query.c:869]: Why are you keep commented the code you are removing?

Comment (erdem2000) [reply web/api/queries/query.c:869]: It is closely related with the logic that I changed and would be useful for better understanding. So personally I would like to keep them until next release. But if it is not the right approach from Netdata coding convention perspective then I can sure remove them.

**Pull request closed without merging**